### ISSUE TITLE
Encoding problem with special characters

### ISSUE THREAD

**Issue by OLChest** [NaturalDocs/NaturalDocs#68]
Hey,

I am currently encountering the issue that I have a bunch of ISO 8859-1 encoded files (about 500) which contain special characters such as `ä`, `ü` and `ö` in them and when using Natural Docs to generate the documentation it displays only a bunch of `?` in the final HTML.

Would be cool if the encoding was configurable and defaults to UTF-8.

**Comment by petrisch** [comment]
I had the same issue, but with essentially Asci files that contained those characters.
My workaround was a script that copied everything over, changed the encoding with iconv (or Set-Content -Encoding UTF8 in Powershell) and then run naturaldocs. If I could configure that in nd would be nice to have I think.

**Comment by NaturalDocs** [comment]
Natural Docs does assume the input files are some variant of UTF (UTF8 with or without a BOM, and UTF16 BE or LE are explicitly unit tested.)  .NET seems to have plenty of support for converting from other encodings, I'd just have to specify them manually when opening the file.

Would the entire input folder always be the same encoding, or would some subfolders be different?  For example, would just supporting this in Project.txt be good enough?

```
Source Folder: C:\MyProject\Source
   Encoding: iso-8859-1
```

or could there potentially be a mix and I'd have to support something like this?

```
Source Folder: C:\MyProject\Source
   Encoding: iso-8859-1 C:\MyProject\Source\x\y\z
   Encoding: iso-8859-1 C:\MyProject\Source\a\b\c
```

Doing it at a file level instead of a folder level isn't likely to be supported.

**Comment by OLChest** [comment]
> Would the entire input folder always be the same encoding, or would some subfolders be different? For example, would just supporting this in Project.txt be good enough? 
> ```
>Source Folder: C:\MyProject\Source
>   Encoding: iso-8859-1
>```

For my use case this would be enough as all code files are always encoded the same. But the option with subfolders looks promising too, as it is configurable and therefore files with other encodings could just be moved into different subfolders.

**Comment by petrisch** [comment]
In my case it would have been something like:
```
Source Folder: C:\MyProject\Source
   Encoding: iso-8859-1 C:\MyProject\Source\x\y\z
   Encoding: iso-8859-1 C:\MyProject\Source\a\b\c
File Filter:
   Encoding: ansi *.mac
   Encoding: utf-8 *.cs
```
But that can have overlappings, so I don't know if its feasible.

**Comment by nassimesi** [comment]
for my case, I am using the 2.1.1 version, but it shows me an error : He doesn't recognise the keywork : "Encoding"

**Comment by OLChest** [comment]
@nassimesi 
> for my case, I am using the 2.1.1 version, but it shows me an error : He doesn't recognise the keywork : "Encoding"

As far as I know, it is not implemented in the version 2.1.1 yet. The feature is being worked on, as commits on the master branch imply.
I tested the current version on the master branch and it works like a charm. But without being marked as stable I would advise against building the solution yourself and wait until the next version with the encoding feature is released and tested thoroughly.

For me it worked just as NaturalDocs mentioned already a few comments above:
```
Source Folder: C:\MyProject\Source
   Encoding: iso-8859-1
```

**Comment by nassimesi** [comment]
@OLChest 
> As far as I know, it is not implemented in the version 2.1.1 yet. The feature is being worked on, as commits on the master branch imply. I tested the current version on the master branch and it works like a charm. But without being marked as stable I would advise against building the solution yourself and wait until the next version with the encoding feature is released and tested thoroughly.
> 
> For me it worked just as NaturalDocs mentioned already a few comments above:
> 
> ```
> Source Folder: C:\MyProject\Source
>    Encoding: iso-8859-1
> ```
How to use the master branch to have an .exe file and launch NaturalDoc ? If there is another way to run NaturalDoc I will be happy to know it

**Comment by OLChest** [comment]
@nassimesi 
> How to use the master branch to have an .exe file and launch NaturalDoc ? If there is another way to run NaturalDoc I will be happy to know it

1. Download Visual Studio 2019 or 2022
2. Clone or download the master branch of this repository
3. Open the `Engine\Engine.csproj` in the repository in Visual Studio
4. Build the `Engine` project by clicking `Build` in the right click menu. This step is necessary because the CLI depends on the Engine and requires its DLL.
5. Add the `CLI\CLI.csproj` to your solution by right clicking on `Solution>Add>Existing Project`
6. Build the `CLI` project by clicking `Build` in the right click menu

Now you should find all the files in `\CLI\bin\Debug Temp`, including the `NaturalDocs.exe`. Make sure the .dll files are always in the same repository as they are needed by NaturalDocs. This is what worked for me.

Again I would not recommend building the application yourself if it ain't for testing only. As long as it ain't an official version there may be some bugs.

**Comment by NaturalDocs** [comment]
This has been implemented in 2.2 Development Release 4.  [How to use it is described here](https://www.naturaldocs.org/news/2021-12-02/Natural_Docs_2.2_development_release_4/), but it's pretty similar to the above.  [Download it here](https://www.naturaldocs.org/download/#Development) and let me know if there are any issues.

**Comment by NaturalDocs** [comment]
[Closing now that 2.2 is released.](https://www.naturaldocs.org/download/)
